File: supabase/functions/_shared/llm-client.ts
/**
 * Single LLM client used by the content pipeline.
 *
 * Configuration comes from Supabase Secrets / env vars only — no
 * hardcoded API keys:
 *
 *   OPENAI_API_KEY
 *   OPENAI_MODEL
 *   LLM_MOCK_MODE
 *
 * When LLM_MOCK_MODE is true (the default), no network call is made and
 * a mock payload with the same JSON shape as a real response is
 * returned, so demos/tests never incur real cost.
 */

import type { LLMPayload } from "./types.ts";
import { parseLlmJson } from "./json-parser.ts";

const DEFAULT_TIMEOUT_MS = 30_000;
const DEFAULT_MODEL = "gpt-4o-mini";
const OPENAI_CHAT_COMPLETIONS_URL = "https://api.openai.com/v1/chat/completions";
const MAX_TOTAL_ATTEMPTS = 2; // one retry, i.e. at most two attempts total

export class LLMError extends Error {}
/** A retryable failure (network error, timeout, 5xx). */
export class TransientLLMError extends LLMError {}

export interface LLMResult {
  content: LLMPayload;
  inputTokens: number;
  outputTokens: number;
  model: string;
}

function isMockMode(): boolean {
  const raw = Deno.env.get("LLM_MOCK_MODE");
  if (raw === undefined) return true;
  return ["1", "true", "yes"].includes(raw.trim().toLowerCase());
}

function getApiKey(): string | null {
  return Deno.env.get("OPENAI_API_KEY") ?? null;
}

function getModel(): string {
  return Deno.env.get("OPENAI_MODEL") ?? DEFAULT_MODEL;
}

function estimateTokens(text: string): number {
  return Math.max(1, Math.floor(text.length / 4));
}

function guessLanguage(text: string): "ru" | "kk" | "en" {
  if (/[әғқңөұүһӘҒҚҢӨҰҮҺ]/.test(text)) return "kk";
  if (/[а-яёА-ЯЁ]/.test(text)) return "ru";
  return "en";
}

function extractSnippet(prompt: string, length = 40): string {
  const marker = 'SOURCE MATERIAL:\n"""\n';
  const idx = prompt.indexOf(marker);
  const raw = idx === -1 ? prompt : prompt.slice(idx + marker.length);
  const snippet = raw.replace(/\s+/g, " ").trim();
  return snippet.slice(0, length) || "Untitled";
}

function buildMockPayload(prompt: string): LLMPayload {
  const language = guessLanguage(prompt);
  const snippet = extractSnippet(prompt);

  const titles: Record<string, string> = {
    ru: `Мок-заголовок: ${snippet}`,
    kk: `Мок-тақырып: ${snippet}`,
    en: `Mock title: ${snippet}`,
  };
  const keyPoints: Record<string, string[]> = {
    ru: [
      "Первый ключевой тезис из материала (мок).",
      "Второй ключевой тезис из материала (мок).",
      "Третий ключевой тезис из материала (мок).",
    ],
    kk: [
      "Материалдың бірінші негізгі тезисі (мок).",
      "Материалдың екінші негізгі тезисі (мок).",
      "Материалдың үшінші негізгі тезисі (мок).",
    ],
    en: [
      "First key takeaway from the material (mock).",
      "Second key takeaway from the material (mock).",
      "Third key takeaway from the material (mock).",
    ],
  };
  const posts: Record<string, { telegram: string; instagram: string; linkedin: string }> = {
    ru: {
      telegram: `[MOCK] ${titles.ru}\n\nКраткий пересказ материала для Telegram.`,
      instagram: `[MOCK] ${titles.ru} ✨ #контент #AI`,
      linkedin: `[MOCK] ${titles.ru} — ключевые выводы для профессионалов.`,
    },
    kk: {
      telegram: `[MOCK] ${titles.kk}\n\nTelegram үшін қысқаша мазмұндама.`,
      instagram: `[MOCK] ${titles.kk} ✨ #контент #AI`,
      linkedin: `[MOCK] ${titles.kk} — мамандарға арналған негізгі қорытындылар.`,
    },
    en: {
      telegram: `[MOCK] ${titles.en}\n\nShort summary of the material for Telegram.`,
      instagram: `[MOCK] ${titles.en} ✨ #content #AI`,
      linkedin: `[MOCK] ${titles.en} — key takeaways for professionals.`,
    },
  };

  return {
    detected_language: language,
    title: titles[language],
    key_points: keyPoints[language],
    telegram: posts[language].telegram,
    instagram: posts[language].instagram,
    linkedin: posts[language].linkedin,
    image_prompt:
      "Abstract conceptual illustration representing the topic, cinematic lighting, " +
      "modern digital art style, no text, no letters, no watermark",
  };
}

function mockGenerate(prompt: string): LLMResult {
  const payload = buildMockPayload(prompt);
  return {
    content: payload,
    inputTokens: estimateTokens(prompt),
    outputTokens: estimateTokens(JSON.stringify(payload)),
    model: "mock",
  };
}

async function requestOpenAi(
  prompt: string,
  apiKey: string,
  model: string,
  fetchImpl: typeof fetch,
): Promise<LLMResult> {
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), DEFAULT_TIMEOUT_MS);

  let response: Response;
  try {
    response = await fetchImpl(OPENAI_CHAT_COMPLETIONS_URL, {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model,
        messages: [{ role: "user", content: prompt }],
        response_format: { type: "json_object" },
        temperature: 0.7,
      }),
      signal: controller.signal,
    });
  } catch (err) {
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new TransientLLMError("Timed out while calling the LLM");
    }
    const message = err instanceof Error ? err.message : String(err);
    throw new TransientLLMError(`Network error while calling the LLM: ${message}`);
  } finally {
    clearTimeout(timeoutId);
  }

  if (!response.ok) {
    if (response.status >= 500) {
      throw new TransientLLMError(`LLM API returned a server error (${response.status})`);
    }
    throw new LLMError(`LLM API returned an error (${response.status})`);
  }

  const data = await response.json();
  const messageContent = data?.choices?.[0]?.message?.content;
  if (typeof messageContent !== "string") {
    throw new LLMError("Unexpected LLM response format: missing message content");
  }

  const parsed = parseLlmJson(messageContent);
  const usage = data?.usage ?? {};

  return {
    content: parsed,
    inputTokens: typeof usage.prompt_tokens === "number" ? usage.prompt_tokens : estimateTokens(prompt),
    outputTokens:
      typeof usage.completion_tokens === "number"
        ? usage.completion_tokens
        : estimateTokens(messageContent),
    model,
  };
}

async function callOpenAiWithRetry(
  prompt: string,
  fetchImpl: typeof fetch,
): Promise<LLMResult> {
  const apiKey = getApiKey();
  if (!apiKey) {
    throw new LLMError("OPENAI_API_KEY is not configured");
  }
  const model = getModel();

  let lastError: unknown;
  for (let attempt = 1; attempt <= MAX_TOTAL_ATTEMPTS; attempt++) {
    try {
      return await requestOpenAi(prompt, apiKey, model, fetchImpl);
    } catch (err) {
      lastError = err;
      const isTransient = err instanceof TransientLLMError;
      if (!isTransient || attempt === MAX_TOTAL_ATTEMPTS) {
        throw err;
      }
    }
  }
  // Unreachable, but keeps TypeScript happy.
  throw lastError instanceof Error ? lastError : new LLMError(String(lastError));
}

/**
 * Generates structured content from the given prompt.
 * `fetchImpl` can be overridden in tests to avoid real network/API calls.
 */
export async function generateJson(
  prompt: string,
  fetchImpl: typeof fetch = fetch,
): Promise<LLMResult> {
  if (isMockMode()) {
    return mockGenerate(prompt);
  }
  return callOpenAiWithRetry(prompt, fetchImpl);
}

File: backend/app/services/llm_client.py
"""Single LLM client used by the content pipeline.

Configuration is read from environment variables only — no hardcoded
API keys:

    GEMINI_API_KEY=
    TEXT_MODEL=gemini-3.1-flash-lite
    LLM_MOCK_MODE=true

When ``LLM_MOCK_MODE`` is true (the default), no network call is made
and a mock payload with the same JSON shape as a real response is
returned. This keeps the test suite free of real, paid API calls.
"""
from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass

import httpx

DEFAULT_TIMEOUT_SECONDS = 30.0
DEFAULT_MODEL = "gemini-3.1-flash-lite"
GEMINI_GENERATE_CONTENT_URL = "https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent"

TRANSIENT_ERRORS = (httpx.TimeoutException, httpx.TransportError)


class LLMError(Exception):
    """Raised when the LLM call fails or returns an unusable response."""


@dataclass
class LLMResult:
    content: dict
    input_tokens: int
    output_tokens: int
    model: str


def _is_mock_mode() -> bool:
    return os.getenv("LLM_MOCK_MODE", "true").strip().lower() in {"1", "true", "yes"}


def _get_api_key() -> str | None:
    return os.getenv("GEMINI_API_KEY") or None


def _get_model() -> str:
    return os.getenv("TEXT_MODEL") or DEFAULT_MODEL


def _estimate_tokens(text: str) -> int:
    return max(1, len(text) // 4)


def _guess_language(text: str) -> str:
    if re.search(r"[әғқңөұүһӘҒҚҢӨҰҮҺ]", text):
        return "kk"
    if re.search(r"[а-яёА-ЯЁ]", text):
        return "ru"
    return "en"


class LLMClient:
    """Thin wrapper around the chat-completions API with retry + mock mode."""

    def __init__(self, timeout: float = DEFAULT_TIMEOUT_SECONDS, max_retries: int = 1):
        self.timeout = timeout
        self.max_retries = max_retries

    def generate_json(self, prompt: str) -> LLMResult:
        if _is_mock_mode():
            return self._mock_generate(prompt)
        return self._call_gemini_with_retry(prompt)

    # -- mock mode -----------------------------------------------------

    def _mock_generate(self, prompt: str) -> LLMResult:
        payload = _build_mock_payload(prompt)
        return LLMResult(
            content=payload,
            input_tokens=_estimate_tokens(prompt),
            output_tokens=_estimate_tokens(json.dumps(payload, ensure_ascii=False)),
            model="mock",
        )

    # -- real mode -------------------------------------------------------

    def _call_gemini_with_retry(self, prompt: str) -> LLMResult:
        api_key = _get_api_key()
        if not api_key:
            raise LLMError("GEMINI_API_KEY is not configured")
        model = _get_model()

        attempts_allowed = self.max_retries + 1
        last_error: Exception | None = None

        for attempt in range(attempts_allowed):
            try:
                return self._request_gemini(prompt, api_key, model)
            except TRANSIENT_ERRORS as exc:
                last_error = exc
                if attempt + 1 < attempts_allowed:
                    continue
                raise LLMError(
                    f"LLM request failed after {attempts_allowed} attempt(s): {exc}"
                ) from exc

        raise LLMError(f"LLM request failed: {last_error}")

    def _request_gemini(self, prompt: str, api_key: str, model: str) -> LLMResult:
        try:
            response = httpx.post(
                GEMINI_GENERATE_CONTENT_URL.format(model=model),
                headers={"x-goog-api-key": api_key},
                json={
                    "contents": [{"role": "user", "parts": [{"text": prompt}]}],
                    "generationConfig": {"responseMimeType": "application/json", "temperature": 0.7},
                },
                timeout=self.timeout,
            )
            response.raise_for_status()
        except httpx.HTTPStatusError as exc:
            raise LLMError(
                f"Gemini API returned an error ({exc.response.status_code})"
            ) from exc

        data = response.json()
        try:
            message_content = data["candidates"][0]["content"]["parts"][0]["text"]
            parsed = json.loads(message_content)
        except (KeyError, IndexError, TypeError, json.JSONDecodeError) as exc:
            raise LLMError(f"Unexpected Gemini response format: {exc}") from exc

        usage = data.get("usageMetadata") or {}
        return LLMResult(
            content=parsed,
            input_tokens=usage.get("promptTokenCount", _estimate_tokens(prompt)),
            output_tokens=usage.get("candidatesTokenCount", _estimate_tokens(message_content)),
            model=model,
        )


def _build_mock_payload(prompt: str) -> dict:
    language = _guess_language(prompt)
    snippet = _extract_snippet(prompt)

    titles = {
        "ru": f"Мок-заголовок: {snippet}",
        "kk": f"Мок-тақырып: {snippet}",
        "en": f"Mock title: {snippet}",
    }
    key_points = {
        "ru": [
            "Первый ключевой тезис из материала (мок).",
            "Второй ключевой тезис из материала (мок).",
            "Третий ключевой тезис из материала (мок).",
        ],
        "kk": [
            "Материалдың бірінші негізгі тезисі (мок).",
            "Материалдың екінші негізгі тезисі (мок).",
            "Материалдың үшінші негізгі тезисі (мок).",
        ],
        "en": [
            "First key takeaway from the material (mock).",
            "Second key takeaway from the material (mock).",
            "Third key takeaway from the material (mock).",
        ],
    }
    posts = {
        "ru": {
            "telegram": f"[MOCK] {titles['ru']}\n\nКраткий пересказ материала для Telegram.",
            "instagram": f"[MOCK] {titles['ru']} ✨ #контент #AI",
            "linkedin": f"[MOCK] {titles['ru']} — ключевые выводы для профессионалов.",
        },
        "kk": {
            "telegram": f"[MOCK] {titles['kk']}\n\nTelegram үшін қысқаша мазмұндама.",
            "instagram": f"[MOCK] {titles['kk']} ✨ #контент #AI",
            "linkedin": f"[MOCK] {titles['kk']} — мамандарға арналған негізгі қорытындылар.",
        },
        "en": {
            "telegram": f"[MOCK] {titles['en']}\n\nShort summary of the material for Telegram.",
            "instagram": f"[MOCK] {titles['en']} ✨ #content #AI",
            "linkedin": f"[MOCK] {titles['en']} — key takeaways for professionals.",
        },
    }

    return {
        "detected_language": language,
        "title": titles[language],
        "key_points": key_points[language],
        "telegram": posts[language]["telegram"],
        "instagram": posts[language]["instagram"],
        "linkedin": posts[language]["linkedin"],
        "image_prompt": (
            "Abstract conceptual illustration representing the topic, cinematic lighting, "
            "modern digital art style, no text, no letters, no watermark"
        ),
    }


def _extract_snippet(prompt: str, length: int = 40) -> str:
    marker = 'SOURCE MATERIAL:\n"""\n'
    idx = prompt.find(marker)
    if idx == -1:
        snippet = prompt.strip()
    else:
        snippet = prompt[idx + len(marker):].strip()
    snippet = re.sub(r"\s+", " ", snippet)
    return snippet[:length].strip() or "Untitled"
